### ISSUE TITLE
Questions about importance weight behaviour

### ISSUE THREAD

**Issue by nmdickson** [joshspeagle/dynesty#339]
Hi there,

I'm trying to use dynesty to fit some model parameters but I'm having a bit of trouble surrounding the transition from the static "initial" batch into the dynamic nested sampling and I was hoping you might be able to provide some insight.

Specifically, I'm running into trouble computing the first set of likelihood bounds for dynamic sampling to start with, due to the shape of the importance weight curve. I've recreated the problem in the plot below using one of the provided toy examples (the eggbox).

![eggbox_jump](https://user-images.githubusercontent.com/63625712/135924271-5f353d3a-204d-4670-89b7-38e0d9f98799.png)

As you can see, when the initial sampling reaches it's stopping condition (dlogz; shown in the legend and indicated by the vertical lines), if it still sits high on the weight curve, adding the remaining live points causes the weight to shoot upwards, often reaching even above the peak.
This then messes up locating the typical set, and leads to nonsensical logl-bounds, which usually stop the dynamic sampling from proceeding.

I think I understand why this is happening, as the log-volume decreases much more quickly when adding in the remaining live points, and so if said points still sit within the typical set and there is lots of volume remaining to cover, their weights will be quite large. Does this intuition sound correct?

The obvious solution would be to decrease dlogz sufficiently so it covers the typical set before stopping (i.e. the red line above),
however in our case we have no desire to estimate the evidence, only our parameter values, and in fact some of our models struggle to reach such low dlogz values. So I was really hoping to use a higher (dlogz~0.5) initial stopping condition if possible.
Do you know if this is at all feasible, or just an unavoidable consequence of using higher dlogz stopping values?

One possible way I noticed for skirting this problem is to switch to using the equally-resampled KDE PDF (`gaussian_kde(resample_equal(-logvol, weights)(-logvol))`) to manually determine the bounds, as that seems to always make the jumps disappear:

![eggbox_jump_pdf](https://user-images.githubusercontent.com/63625712/135928228-8e1a77c6-d339-44ac-a235-67dff9279a74.png)

Although I don't understand why and am thus a bit wary of it. Do you think this might still be a safe approach?

Thanks very much,
Nolan

**Comment by segasai** [comment]
Hi Nolan, 

Are you using the version from github or 1.1 ? There were some changes (improvements) in how dynamic sampling proceeds, how new bounds are selected and when dynamic sampling stops. 
Also if you don't care about logz at you should set stop_kwargs={pfrac:1} and wt_kwargs={pfrac:1} as that will ensure dynamic sampling will only keep the posterior in mind when allocating batches/stopping.

It is also not obvious to me why any of those weight curves are problematic at all, i.e. I don't see jumps as an issue. 

Also if I understand correctly for a N-dimensional Gaussian the midpoint of th e typical set is located at dlogz=0.5 or so, but the actual points in the typical set will extend to dlogz=0.01 easily, so it's not clear to me that dlogz=0.01 is a bad default.

**Comment by nmdickson** [comment]
> Are you using the version from github or 1.1 ? There were some changes (improvements) in how dynamic sampling proceeds, how new bounds are selected and when dynamic sampling stops.

I'm using the latest version from github (and I definitely noticed improvements when I switched to it).

> Also if you don't care about logz at you should set stop_kwargs={pfrac:1} and wt_kwargs={pfrac:1} as that will ensure dynamic sampling will only keep the posterior in mind when allocating batches/stopping.
> 
> It is also not obvious to me why any of those weight curves are problematic at all, i.e. I don't see jumps as an issue.

I do have `pfrac` always set to 1, the problem occurs sometimes when the default `dynamicsampler.weight_function` tries to compute the first set of likelihood bounds (at 80% of the max pweight). In the more extreme cases I've seen this can miss the typical set completely. Take, for example, this run of my actual model:

![model_jump](https://user-images.githubusercontent.com/63625712/136038634-56b4e1b0-dfe3-4d98-b019-a93fe68dae19.png)

This, in the newest versions of the code, chooses the logl bounds to be from a point or two to the left of the max, and infinity on the right. This usually gives a very high logl for the minimum bound (which is hard to sample greater than) and also misses most of the typical set, I think.

> Also if I understand correctly for a N-dimensional Gaussian the midpoint of th e typical set is located at dlogz=0.5 or so, but the actual points in the typical set will extend to dlogz=0.01 easily, so it's not clear to me that dlogz=0.01 is a bad default.

Ideally I would just use the default conditions and not have to worry about all this, but some of our actual models struggle quite a bit to reach the lowest dlogz values, and spend an inordinate amount of time trying to sample the highest likelihoods possible rather than just the typical set, which is what we really care about. So I was hoping I could cut this part off early and just jump into the dynamic sampling, which works very well if the initial logl bounds surround the typical set.
(Why some of our models take so long near the very peak of the posterior is a whole other problem I'm looking into)

**Comment by segasai** [comment]
To be honest, what you describe and plot looks totally normal. I.e. You set-up dynamic sampling that you decide to cut a bit short, just after you've passed the typical set. Because now you have essentially cut-off the tail of the distribution that is being sampled by only Nlive points (and with *uniform* volume sampling as opposed to log-uniform), those will have higher weights. Since the dynamic sampler 'aims' at weight equality, it'll focus the first sampling at those high weight regions. 

I think it is perfectly possible that there is some inefficiency and/or bug somewhere, but at the moment I'm not seeing it. If you have some runnable test-case or demonstration on the egg-shell or whatever, that may help make some progress.

**Comment by nmdickson** [comment]
Sorry, I don't mean to imply there's an issue with how dynesty handles this, I agree the math makes sense and in general this is how the sampler should behave.

I was more just wondering if, because I only wish to focus on the typical set, it's safe for me to ignore this jump and "direct" the dynamic sampler specifically at the typical set (which I think the KDE method from above would accomplish), instead of simply the highest weighted region.
In other words, are those final points (and dynamic sampling around them)  _really_ important for say, parameter estimation, like their weights might suggest, or can I focus the rest of the sampling directly at the typical set.

**Comment by nmdickson** [comment]
> Based on how I understand the problem, if the weights that you show in the plot are correct, I think it is sensible behaviour for the sampler to go there if you care about parameter estimation. The reason for this is if you care about the effective number of samples I.e. Neff = 1./\sum w_i^2 and you are trying to identify where to add points, the point with the highest weights are the points where you'll be most efficient in increasing Neff (under some assumptions). And I think increasing neff is what you should care about.

That makes sense, in terms of Neff and weights.
It just seemed to me like these "jumps" are a bit artificial, if you compared say the blue against red weights in the first plot above, and asked me to define where the "typical set" is on that plot.
But I think maybe that's just a misinterpretation of the weights-volume graph on my part, and in reality, like you said, those jumps aren't really in the "tails", but rather are an integral part of the "bulk" of the typical set, and probably shouldn't just be ignored.

> Also, I was previously thinking about where to allocate new batches, and it's clear that weights > .8*weight.max() is only approximately correct in some ideal situations, maybe there is a more justified strategy.

I think this is also the source of my concerns. Using this method taking 80% of the max weight sometimes (as shown in the third figure above) misses the majority of the typical set completely, which seemed troubling to me.
I know simply using a smaller fraction might work in my case, but that doesn't seem like a very general solution. 

I'm not sure if there might be a better solution involving finding any major peaks in the weights and making sure you cover them all or something. I don't know, I don't think I'm best informed to solve that problem.

> Regarding the KDE approach, I'm not sure I fully understood what you propose...

I'm less sure this is a good idea now, if these jumps shouldn't be ignored, but just for posterity, what I was talking about was I guess a "smoothing" of the weights curve by using the `resample_equal` and `gaussian_kde` functions, like is done in some of dynesty's plots. This causes the "jumps" to disappear completely, in all cases.
See this gist for how it works on the eggbox demo: https://gist.github.com/nmdickson/1645aa5f30a147818f550de84178630a
I'm not entirely sure why it works the way it does, which is why I was reluctant to use it in the first place.
